Python/errors.c

```diff
@@ -1850,6 +1850,54 @@ PyErr_SyntaxLocationEx(const char *filename, int lineno, int col_offset)
     Py_XDECREF(fileobj);
 }
 
+/* Raises a SyntaxError and returns -1.
+ * If something goes wrong, a different exception may be raised.
+ */
+int
+_PyErr_RaiseSyntaxError(PyObject *msg, PyObject *filename, int lineno, int col_offset,
+                        int end_lineno, int end_col_offset)
+{
+    PyObject *loc_obj = PyErr_ProgramTextObject(filename, lineno);
```

Comment:
Perhaps `line` or like be better bane than `loc_obj`.

Reply:
`PyErr_ProgramTextObject()` always uses UTF-8 to decode the line. This may be wrong for Python source. You should use `_PyErr_ProgramDecodedTextObject()` with the encoding of the Python source.

Reply:
> Perhaps `line` or like be better bane than `loc_obj`.

it's called `text` in the SyntaxError definition.

Reply:
> `PyErr_ProgramTextObject()` always uses UTF-8 to decode the line. This may be wrong for Python source. You should use `_PyErr_ProgramDecodedTextObject()` with the encoding of the Python source.

Let's do this in a separate PR, with tests.

Reply:
There may even be already opened issue for this.

In any case, this  means that you need to pass _encoding_ in this API.

Reply:
This is why I want to do it separately - I don't have the encoding in the compiler at the moment.

It seems that this function is only ever called with NULL for encoding at the moment:

```
% grep -rI _PyErr_ProgramDecodedTextObject Modules Objects Python 
Python/errors.c:_PyErr_ProgramDecodedTextObject(PyObject *filename, int lineno, const char* encoding)
Python/errors.c:    return _PyErr_ProgramDecodedTextObject(filename, lineno, NULL);
```

Reply:
See #124188.

`_PyErr_ProgramDecodedTextObject()` is used in the parser to pass the encoding which is known at that moment. Some new users of `_PyErr_RaiseSyntaxError()` also can know the source encoding.